File: Problems/No.47_Permutations_II/permutations_II.cpp
class Solution {
	private:
        
        // find all permutations using backtracking
        void solve(int idx,vector<int> nums,vector<vector<int>>& ans){

            if(idx == nums.size()){

                ans.push_back(nums);
                return;

            }
            else {
                
                for(int i = idx; i < nums.size(); i++){

                    //avoid duplicates
                    if(i!=idx && nums[i]==nums[idx]) continue;
                    else {

                        swap(nums[i], nums[idx]);
                        solve(idx+1, nums, ans);

                    }

                }                
                
            }

        }

	public:
    
		vector<vector<int>> permuteUnique(vector<int>& nums) {
            
            // parameter
			vector<vector<int>> all_permutations;
            
            // start compute, using backtracking
			sort(nums.begin(),nums.end());
			solve(0, nums, all_permutations);
            
			return all_permutations;
            
		}
};

File: Problems/No.47_Permutations_II/permutations_II.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        
        # find all permutations using backtracking
        def solve(nums, permutation):
            # check if it duplicates
            if ((len(permutation)==length) and (permutation not in all_permutations)):
                all_permutations.append(permutation)
            for i in range(len(nums)):
                solve(nums[:i]+nums[i+1:], permutation + [nums[i]])
                
        # parameters
        all_permutations = []  
        length = len(nums)
        
        # start compute
        nums.sort()
        solve (nums, [])
        return all_permutations
